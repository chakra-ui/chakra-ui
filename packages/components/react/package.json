{
  "name": "@chakra-ui/react",
  "version": "2.4.9",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "src/index.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chakra-ui/accordion": "workspace:*",
    "@chakra-ui/alert": "workspace:*",
    "@chakra-ui/avatar": "workspace:*",
    "@chakra-ui/breadcrumb": "workspace:*",
    "@chakra-ui/button": "workspace:*",
    "@chakra-ui/card": "workspace:*",
    "@chakra-ui/checkbox": "workspace:*",
    "@chakra-ui/close-button": "workspace:*",
    "@chakra-ui/control-box": "workspace:*",
    "@chakra-ui/counter": "workspace:*",
    "@chakra-ui/css-reset": "workspace:*",
    "@chakra-ui/editable": "workspace:*",
    "@chakra-ui/form-control": "workspace:*",
    "@chakra-ui/hooks": "workspace:*",
    "@chakra-ui/icon": "workspace:*",
    "@chakra-ui/image": "workspace:*",
    "@chakra-ui/input": "workspace:*",
    "@chakra-ui/layout": "workspace:*",
    "@chakra-ui/live-region": "workspace:*",
    "@chakra-ui/media-query": "workspace:*",
    "@chakra-ui/menu": "workspace:*",
    "@chakra-ui/modal": "workspace:*",
    "@chakra-ui/number-input": "workspace:*",
    "@chakra-ui/pin-input": "workspace:*",
    "@chakra-ui/popover": "workspace:*",
    "@chakra-ui/popper": "workspace:*",
    "@chakra-ui/portal": "workspace:*",
    "@chakra-ui/progress": "workspace:*",
    "@chakra-ui/provider": "workspace:*",
    "@chakra-ui/radio": "workspace:*",
    "@chakra-ui/react-env": "workspace:*",
    "@chakra-ui/select": "workspace:*",
    "@chakra-ui/skeleton": "workspace:*",
    "@chakra-ui/slider": "workspace:*",
    "@chakra-ui/spinner": "workspace:*",
    "@chakra-ui/stat": "workspace:*",
    "@chakra-ui/styled-system": "workspace:*",
    "@chakra-ui/switch": "workspace:*",
    "@chakra-ui/system": "workspace:*",
    "@chakra-ui/table": "workspace:*",
    "@chakra-ui/tabs": "workspace:*",
    "@chakra-ui/tag": "workspace:*",
    "@chakra-ui/textarea": "workspace:*",
    "@chakra-ui/theme": "workspace:*",
    "@chakra-ui/theme-utils": "workspace:*",
    "@chakra-ui/toast": "workspace:*",
    "@chakra-ui/tooltip": "workspace:*",
    "@chakra-ui/transition": "workspace:*",
    "@chakra-ui/utils": "workspace:*",
    "@chakra-ui/visually-hidden": "workspace:*"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": ">=4.0.0",
    "react": ">=18",
    "react-dom": ">=18"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/components/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://storybook.chakra-ui.com"
  },
  "devDependencies": {
    "@emotion/react": "^11.9.0",
    "@emotion/styled": "^11.8.1",
    "framer-motion": "^6.2.9",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "clean-package": "2.1.1"
  },
  "scripts": {
    "build": "tsup src --dts",
    "dev": "pnpm build:fast -- --watch",
    "clean": "rimraf dist .turbo",
    "typecheck": "tsc --noEmit",
    "build:fast": "tsup src",
    "prepack": "clean-package",
    "postpack": "clean-package restore"
  },
  "clean-package": "../../../clean-package.config.json",
  "tsup": {
    "clean": true,
    "target": "es2019",
    "format": [
      "cjs",
      "esm"
    ]
  }
}
