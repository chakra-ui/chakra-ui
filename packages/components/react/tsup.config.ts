import fs from "node:fs"
import path from "node:path"
import { defineConfig } from "tsup"

function removeDirective(fileName: string) {
  const file = path.resolve(__dirname, "dist", fileName)
  return fs.promises.writeFile(
    file,
    fs.readFileSync(file, "utf-8").replace("'use client'\n", ""),
  )
}

export default defineConfig({
  format: ["cjs", "esm"],
  target: "es2019",
  sourcemap: true,
  banner: {
    js: "'use client'",
  },
  async onSuccess() {
    await Promise.all([
      removeDirective("index.mjs"),
      removeDirective("index.js"),
    ])
  },
})
