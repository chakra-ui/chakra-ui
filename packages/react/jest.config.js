const baseConfig = require("../../jest.config")

module.exports = {
  ...baseConfig,
}
