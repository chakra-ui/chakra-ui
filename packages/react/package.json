{
  "name": "@chakra-ui/react",
  "version": "1.6.12",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/types/index.d.ts",
  "typings": "dist/types/index.d.ts",
  "exports": {
    ".": {
      "require": "./dist/cjs/index.js",
      "default": "./dist/esm/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "start": "nodemon --watch src --exec yarn build -e ts,tsx",
    "build": "concurrently yarn:build:*",
    "test": "jest --env=jsdom --passWithNoTests",
    "lint": "concurrently yarn:lint:*",
    "version": "yarn build",
    "build:esm": "cross-env BABEL_ENV=esm babel src --root-mode upward --extensions .ts,.tsx -d dist/esm --source-maps",
    "build:cjs": "cross-env BABEL_ENV=cjs babel src --root-mode upward --extensions .ts,.tsx -d dist/cjs --source-maps",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationDir dist/types",
    "test:cov": "yarn test --coverage",
    "lint:src": "eslint src --ext .ts,.tsx --config ../../.eslintrc",
    "lint:types": "tsc --noEmit"
  },
  "dependencies": {
    "@chakra-ui/accordion": "1.3.10",
    "@chakra-ui/alert": "1.2.9",
    "@chakra-ui/avatar": "1.2.12",
    "@chakra-ui/breadcrumb": "1.2.10",
    "@chakra-ui/button": "1.4.6",
    "@chakra-ui/checkbox": "1.5.10",
    "@chakra-ui/close-button": "1.1.13",
    "@chakra-ui/control-box": "1.0.17",
    "@chakra-ui/counter": "1.1.11",
    "@chakra-ui/css-reset": "1.0.0",
    "@chakra-ui/editable": "1.2.11",
    "@chakra-ui/form-control": "1.4.3",
    "@chakra-ui/hooks": "1.6.2",
    "@chakra-ui/icon": "1.1.13",
    "@chakra-ui/image": "1.0.22",
    "@chakra-ui/input": "1.2.12",
    "@chakra-ui/layout": "1.4.11",
    "@chakra-ui/live-region": "1.0.16",
    "@chakra-ui/media-query": "1.1.5",
    "@chakra-ui/menu": "1.7.8",
    "@chakra-ui/modal": "1.9.4",
    "@chakra-ui/number-input": "1.2.12",
    "@chakra-ui/pin-input": "1.6.7",
    "@chakra-ui/popover": "1.9.1",
    "@chakra-ui/popper": "2.3.1",
    "@chakra-ui/portal": "1.2.11",
    "@chakra-ui/progress": "1.1.16",
    "@chakra-ui/provider": "1.6.11",
    "@chakra-ui/radio": "1.3.13",
    "@chakra-ui/react-env": "1.0.8",
    "@chakra-ui/select": "1.1.17",
    "@chakra-ui/skeleton": "1.1.21",
    "@chakra-ui/slider": "1.4.2",
    "@chakra-ui/spinner": "1.1.14",
    "@chakra-ui/stat": "1.1.14",
    "@chakra-ui/switch": "1.2.13",
    "@chakra-ui/system": "1.7.6",
    "@chakra-ui/table": "1.2.8",
    "@chakra-ui/tabs": "1.5.7",
    "@chakra-ui/tag": "1.1.14",
    "@chakra-ui/textarea": "1.1.16",
    "@chakra-ui/theme": "1.11.1",
    "@chakra-ui/toast": "1.3.4",
    "@chakra-ui/tooltip": "1.3.14",
    "@chakra-ui/transition": "1.3.8",
    "@chakra-ui/utils": "1.8.4",
    "@chakra-ui/visually-hidden": "1.0.16"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "3.x || 4.x || 5.x",
    "react": ">=16.8.6",
    "react-dom": ">=16.8.6"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://chakra-ui.netlify.app"
  },
  "devDependencies": {
    "@emotion/react": "^11.1.4",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "^4.0.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.1"
  }
}
