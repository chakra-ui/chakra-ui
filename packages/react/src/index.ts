export * from "./collection"
export * from "./components"
export { createContext } from "./create-context"
export * from "./hooks"
export { mergeProps } from "./merge-props"
export { mergeRefs } from "./merge-refs"
export { defaultConfig, defaultSystem } from "./preset"
export { defaultBaseConfig } from "./preset-base"
export * from "./styled-system"
