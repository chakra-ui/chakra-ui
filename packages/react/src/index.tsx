export * from "./chakra-provider"
export * from "./extend-theme"
export * from "./theme-extensions"
export * from "@chakra-ui/accordion"
export * from "@chakra-ui/alert"
export * from "@chakra-ui/avatar"
export * from "@chakra-ui/breadcrumb"
export * from "@chakra-ui/button"
export * from "@chakra-ui/checkbox"
export * from "@chakra-ui/close-button"
export * from "@chakra-ui/counter"
export * from "@chakra-ui/css-reset"
export * from "@chakra-ui/editable"
export * from "@chakra-ui/form-control"
export * from "@chakra-ui/control-box"
export * from "@chakra-ui/hooks"
export * from "@chakra-ui/icon"
export * from "@chakra-ui/image"
export * from "@chakra-ui/input"
export * from "@chakra-ui/layout"
export type { GridProps, ListProps } from "@chakra-ui/layout"
export * from "@chakra-ui/media-query"
export * from "@chakra-ui/table"
export * from "@chakra-ui/theme"
export * from "@chakra-ui/menu"
export * from "@chakra-ui/modal"
export * from "@chakra-ui/number-input"
export * from "@chakra-ui/pin-input"
export * from "@chakra-ui/popover"
export * from "@chakra-ui/popper"
export * from "@chakra-ui/portal"
export * from "@chakra-ui/progress"
export * from "@chakra-ui/radio"
export * from "@chakra-ui/select"
export * from "@chakra-ui/skeleton"
export * from "@chakra-ui/slider"
export * from "@chakra-ui/spinner"
export * from "@chakra-ui/stat"
export * from "@chakra-ui/switch"
export * from "@chakra-ui/system"
export * from "@chakra-ui/tabs"
export * from "@chakra-ui/tag"
export * from "@chakra-ui/textarea"
export * from "@chakra-ui/toast"
export * from "@chakra-ui/tooltip"
export * from "@chakra-ui/transition"
export * from "@chakra-ui/visually-hidden"
