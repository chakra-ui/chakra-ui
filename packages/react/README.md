# Welcome to Chakra UI ⚡️

[![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg?style=flat-square)](#contributors-)

- Works out of the box. Chakra UI contains a set of polished React components
  that work out of the box.

- Flexible & composable. Chakra UI components are built on top of a React UI
  Primitive for endless composability.

- Accessible. Chakra UI components follows the WAI-ARIA guidelines
  specifications.

- Dark Mode 😍: All components are dark mode compatible.

## Looking for the documentation?

https://chakra-ui.com

## Installing Chakra UI

⚡️Chakra UI is made up of multiple components and tools which you can import
one by one. All you need to do is install the `@chakra-ui/react` package:

```sh
$ yarn add @chakra-ui/react
# or
$ npm install --save @chakra-ui/react
```

# Getting set up

To start using the components, please follow these steps:

1. Wrap your application in a `ThemeProvider` provided by **@chakra-ui/react**

```jsx
import { ThemeProvider, ColorModeProvider } from "@chakra-ui/react"

const App = ({ children }) => (
  <ThemeProvider>
    <ColorModeProvider>{children}</ColorModeProvider>
  </ThemeProvider>
)
```

`ColorModeProvider` is a context that provides light mode and dark mode values
to the components. It also comes with a function to toggle between light/dark
mode.

2. Now you can start using components like so!:

```jsx
import { Button } from "@chakra-ui/react"

const App = () => <Button>I just consumed some ⚡️Chakra!</Button>
```

# Contributing

Feel like contributing? That's awesome! We have a
[contributing guide](../../CONTRIBUTING.md) to help guide you.

The components to be built come from the
[Aria Practices Design Patterns and Widgets](https://www.w3.org/TR/wai-aria-practices-1.1).

## Contributors ✨

Thanks goes to these wonderful people
([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/segunadebayo"><img src="https://avatars2.githubusercontent.com/u/6916170?v=4" width="100px;" alt="Segun Adebayo"/><br /><sub><b>Segun Adebayo</b></sub></a><br /><a href="https://github.com/chakra-ui/chakra-ui/commits?author=segunadebayo" title="Code">💻</a> <a href="#maintenance-segunadebayo" title="Maintenance">🚧</a> <a href="https://github.com/chakra-ui/chakra-ui/commits?author=segunadebayo" title="Documentation">📖</a> <a href="#example-segunadebayo" title="Examples">💡</a> <a href="#design-segunadebayo" title="Design">🎨</a></td>
    <td align="center"><a href="https://github.com/tioluwani94"><img src="https://avatars1.githubusercontent.com/u/11310046?v=4" width="100px;" alt="Tioluwani Kolawole"/><br /><sub><b>Tioluwani Kolawole</b></sub></a><br /><a href="https://github.com/chakra-ui/chakra-ui/commits?author=tioluwani94" title="Documentation">📖</a> <a href="#example-tioluwani94" title="Examples">💡</a> <a href="#maintenance-tioluwani94" title="Maintenance">🚧</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the
[all-contributors](https://github.com/all-contributors/all-contributors)
specification. Contributions of any kind welcome!
