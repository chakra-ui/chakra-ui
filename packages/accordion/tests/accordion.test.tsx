import * as React from "react"
import {
  userEvent,
  render,
  fireEvent,
  testA11y,
  screen,
} from "@chakra-ui/test-utils"
import {
  Accordion,
  AccordionButton,
  AccordionItem,
  AccordionPanel,
} from "../src"

jest.mock("@chakra-ui/collapse", () => {
  const Collapse = jest.fn(({ children, isOpen }) => (
    <div id="collapse-mock" hidden={!isOpen}>
      {children}
    </div>
  ))
  return { Collapse }
})

it("passes a11y test", async () => {
  await testA11y(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
})

test("uncontrolled: It opens the accordion panel", () => {
  render(
    <Accordion defaultIndex={0}>
      <AccordionItem>
        <AccordionButton data-testid="button">Section 1 title</AccordionButton>
        <AccordionPanel data-testid="panel">Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  expect(screen.getByTestId("button")).toHaveAttribute("aria-expanded", "true")
})

test("uncontrolled: toggles the accordion on click", async () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Trigger</AccordionButton>
        <AccordionPanel>Panel</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  const trigger = screen.getByText("Trigger")

  userEvent.click(trigger)
  expect(trigger).toHaveAttribute("aria-expanded", "true")

  // you can't toggle an accordion without passing `allowToggle`
  userEvent.click(trigger)
  expect(trigger).toHaveAttribute("aria-expanded", "true")
})

// test that arrow up & down moves focus to next/previous accordion
test("arrow up & down moves focus to next/previous accordion", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Section 2 title</AccordionButton>
        <AccordionPanel>Panel 2</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const first = screen.getByText("Section 1 title")
  const second = screen.getByText("Section 2 title")

  fireEvent.keyDown(first, { key: "ArrowDown", keyCode: 40 })
  expect(second).toHaveFocus()

  fireEvent.keyDown(second, { key: "ArrowUp", keyCode: 38 })
  expect(first).toHaveFocus()
})

// test that home & end keys moves focus to first/last accordion
test("home & end keys moves focus to first/last accordion", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>First section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Second section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Last section</AccordionButton>
        <AccordionPanel>Panel 2</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const first = screen.getByText("First section")
  const last = screen.getByText("Last section")

  fireEvent.keyDown(first, { key: "Home", keyCode: 36 })
  expect(first).toHaveFocus()

  fireEvent.keyDown(first, { key: "End", keyCode: 35 })
  expect(last).toHaveFocus()
})

// test the only one accordion can be visible + is not togglable
test("only one accordion can be visible + is not togglable", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>First section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Second section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  const firstAccordion = screen.getByText("First section")

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "true")

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "true")
})
// test the only one accordion can be visible + is togglable
test("only one accordion can be visible + is togglable", () => {
  render(
    <Accordion allowToggle>
      <AccordionItem>
        <AccordionButton>First section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Second section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  const firstAccordion = screen.getByText("First section")

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "true")

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "false")
})

// test that multiple accordions can be opened + is togglable
test("multiple accordions can be opened + is togglable", () => {
  render(
    <Accordion allowMultiple>
      <AccordionItem>
        <AccordionButton>First section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Second section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  const firstAccordion = screen.getByText("First section")
  const secondAccordion = screen.getByText("Second section")

  userEvent.click(firstAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "true")

  userEvent.click(secondAccordion)
  expect(firstAccordion).toHaveAttribute("aria-expanded", "true")
})

// it has the proper aria attributes
test("has the proper aria attributes", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const button = screen.getByText("Section 1 title")
  const panel = screen.getByText("Panel 1")

  expect(button).toHaveAttribute("aria-controls")
  expect(button).toHaveAttribute("aria-expanded")
  expect(panel).toHaveAttribute("aria-labelledby")
})

// test that tab moves focus to the next focusable element
test("tab moves focus to the next focusable element", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>First section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Second section</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>

      <AccordionItem>
        <AccordionButton>Last section</AccordionButton>
        <AccordionPanel>Panel 2</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const first = screen.getByText("First section")
  const second = screen.getByText("Second section")
  const last = screen.getByText("Last section")

  userEvent.tab()
  expect(first).toHaveFocus()

  userEvent.tab()
  expect(second).toHaveFocus()

  userEvent.tab()
  expect(last).toHaveFocus()
})

// test that aria-contols for button is same as id for panel
test("aria-contols for button is same as id for panel", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const button = screen.getByText("Section 1 title")
  const panel = screen.getByText("Panel 1")
  expect(button.getAttribute("aria-controls")).toEqual(panel.getAttribute("id"))
})

// test that aria-expanded is true/false when accordion is open/closed
test("aria-expanded is true/false when accordion is open/closed", () => {
  render(
    <Accordion defaultIndex={0}>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
      <AccordionItem>
        <AccordionButton>Section 2 title</AccordionButton>
        <AccordionPanel>Panel 2</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )

  const button = screen.getByText("Section 1 title")
  expect(button).toHaveAttribute("aria-expanded", "true")
})

// test that panel has role=region and aria-labelledby
test("panel has role=region and aria-labelledby", () => {
  render(
    <Accordion>
      <AccordionItem>
        <AccordionButton>Section 1 title</AccordionButton>
        <AccordionPanel>Panel 1</AccordionPanel>
      </AccordionItem>
    </Accordion>,
  )
  const panel = screen.getByText("Panel 1")

  expect(panel).toHaveAttribute("aria-labelledby")
  expect(panel).toHaveAttribute("role", "region")
})
