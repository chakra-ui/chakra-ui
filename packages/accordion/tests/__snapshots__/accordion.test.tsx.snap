// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accordion renders correctly 1`] = `
<DocumentFragment>
  .emotion-2 {
  border-top-width: 1px;
  border-color: inherit;
}

.emotion-2:last-of-type {
  border-bottom-width: 1px;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  outline: 0;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
}

.emotion-0:focus,
.emotion-0[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: rgba(0,0,0,0.04);
}

.emotion-0[disabled],
.emotion-0[disabled]:focus,
.emotion-0[disabled]:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover,
.emotion-0[data-disabled],
.emotion-0[data-disabled]:focus,
.emotion-0[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-1 {
  padding-top: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-bottom: 1.25rem;
  -webkit-transition: height 150ms ease-in-out,opacity 150ms ease-in-out,-webkit-transform 150ms ease-in-out;
  -webkit-transition: height 150ms ease-in-out,opacity 150ms ease-in-out,transform 150ms ease-in-out;
  transition: height 150ms ease-in-out,opacity 150ms ease-in-out,transform 150ms ease-in-out;
}

<div
    class="chakra-accordion emotion-3"
  >
    <div
      class="chakra-accordion__item emotion-2"
    >
      <button
        aria-controls="accordion-panel-1"
        aria-expanded="false"
        class="chakra-accordion__button emotion-0"
        id="accordion-button-1"
      >
        Section 1 title
      </button>
      <div
        hidden=""
        id="collapse-mock"
      >
        <div
          aria-labelledby="accordion-button-1"
          class="chakra-accordion__panel emotion-1"
          id="accordion-panel-1"
          role="region"
        >
          Panel 1
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
